File: backend/app/services/coach_service.py
"""AI dating coach service powered by OpenAI GPT."""

from __future__ import annotations

import logging

from openai import AsyncOpenAI

from app.config import get_settings

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = """You are GigaRizz Coach, an expert AI dating coach that helps people create 
amazing dating profiles and have better conversations on dating apps. You are:

- Supportive and encouraging, never judgmental
- Practical with actionable advice backed by dating app data
- Aware of platform differences (Tinder, Hinge, Bumble, Raya)
- Skilled at writing bios, openers, and conversation replies
- Focused on authenticity — help users be their BEST selves, not fake

Rules:
- Keep bios under 500 characters unless specified
- Opening lines should be unique, not generic copy-paste
- Never be creepy, pushy, or disrespectful
- Adapt tone to what the user asks for (witty/sincere/bold)
- Reference specific details when context is provided
"""


class CoachService:
    """Handles AI dating coach interactions via OpenAI."""

    def __init__(self):
        settings = get_settings()
        self.client = AsyncOpenAI(api_key=settings.openai_api_key)
        self.model = settings.openai_model

    async def generate_bio(
        self,
        interests: list[str],
        tone: str,
        platform: str,
        age: int | None = None,
        gender: str | None = None,
    ) -> dict:
        """Generate a dating bio with alternatives."""
        user_context = f"Interests: {', '.join(interests)}"
        if age:
            user_context += f"\nAge: {age}"
        if gender:
            user_context += f"\nGender: {gender}"

        prompt = f"""Write a {tone} dating bio for {platform}. 

{user_context}

Requirements:
- Under 500 characters for the main bio
- Authentic and specific to their interests
- Optimized for {platform}'s format and audience
- Include a hook, personality reveal, and soft CTA

Return exactly 3 bio options, separated by ---"""

        response = await self.client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
            temperature=0.9,
            max_tokens=1000,
        )

        content = response.choices[0].message.content or ""
        bios = [b.strip() for b in content.split("---") if b.strip()]

        return {
            "bio": bios[0] if bios else content,
            "alternatives": bios[1:] if len(bios) > 1 else [],
            "character_count": len(bios[0]) if bios else len(content),
        }

    async def generate_openers(
        self,
        match_name: str,
        platform: str,
        context: str | None = None,
        count: int = 3,
    ) -> list[str]:
        """Generate conversation opening lines."""
        context_str = f"About them: {context}" if context else "No specific context provided."

        prompt = f"""Write {count} unique opening messages for {match_name} on {platform}.

{context_str}

Requirements:
- Each opener should have a different approach (funny, curious, direct)
- Reference their profile details if context is provided
- Keep each under 200 characters
- Be engaging, not generic
- No pickup lines or anything cringe

Return each opener on a new line, numbered 1-{count}."""

        response = await self.client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
            temperature=0.95,
            max_tokens=600,
        )

        content = response.choices[0].message.content or ""
        # Parse numbered lines
        lines = []
        for line in content.strip().split("\n"):
            cleaned = line.strip()
            # Remove numbering like "1. " or "1) "
            if cleaned and cleaned[0].isdigit():
                cleaned = cleaned.lstrip("0123456789.)- ").strip()
            if cleaned:
                lines.append(cleaned)
        return lines[:count]

    async def generate_hinge_prompts(self) -> list[dict]:
        """Generate Hinge prompt + answer pairs."""
        prompt = """Write 3 Hinge prompt answers. Use real Hinge prompts.

Format each as:
PROMPT: [the hinge prompt]
ANSWER: [a great answer]

Requirements:
- Answers should be witty, specific, and conversation-starting
- Mix humor with genuine personality
- Keep answers under 150 characters each
- Use prompts that actually exist on Hinge"""

        response = await self.client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
            temperature=0.9,
            max_tokens=600,
        )

        content = response.choices[0].message.content or ""
        results = []
        current_prompt = ""
        for line in content.strip().split("\n"):
            line = line.strip()
            if line.upper().startswith("PROMPT:"):
                current_prompt = line[7:].strip()
            elif line.upper().startswith("ANSWER:") and current_prompt:
                results.append({"prompt": current_prompt, "answer": line[7:].strip()})
                current_prompt = ""

        return results or [
            {"prompt": "The way to win me over is", "answer": "Show up with curiosity and good restaurant picks"},
            {"prompt": "I'm looking for", "answer": "Someone who matches my energy at concerts AND on lazy Sundays"},
            {"prompt": "A life goal of mine", "answer": "Visit every continent with a local friend on each one"},
        ]

    async def suggest_replies(
        self,
        message: str,
        match_name: str,
        conversation_context: str | None = None,
        count: int = 3,
    ) -> list[str]:
        """Suggest conversation replies."""
        context_str = f"Conversation so far: {conversation_context}" if conversation_context else ""

        prompt = f"""Suggest {count} reply options to this message from {match_name}:

"{message}"

{context_str}

Requirements:
- Each reply should have a different energy (playful, interested, flirty)
- Keep each under 200 characters
- Move the conversation forward
- Be authentic, not try-hard

Return each reply on a new line, numbered 1-{count}."""

        response = await self.client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
            temperature=0.9,
            max_tokens=500,
        )

        content = response.choices[0].message.content or ""
        lines = []
        for line in content.strip().split("\n"):
            cleaned = line.strip()
            if cleaned and cleaned[0].isdigit():
                cleaned = cleaned.lstrip("0123456789.)- ").strip()
            if cleaned:
                lines.append(cleaned)
        return lines[:count]

    async def close(self):
        await self.client.close()

File: GigaRizz/Sources/Core/Services/CoachService.swift
import Foundation

// MARK: - Coach Service

/// AI-powered dating coach service for bios, openers, and conversation tips.
/// Uses ServiceMode to switch between mock (local templates) and production (OpenAI via Cloud Functions).
@MainActor
final class CoachService: ObservableObject {
    // MARK: - Singleton

    static let shared = CoachService()

    // MARK: - Published Properties

    @Published var isLoading = false
    @Published var errorMessage: String?

    // MARK: - Chat Message

    struct ChatMessage: Identifiable, Equatable {
        let id: String
        let role: Role
        let content: String
        let timestamp: Date

        enum Role: String {
            case user, assistant, system
        }

        init(
            id: String = UUID().uuidString,
            role: Role,
            content: String,
            timestamp: Date = Date()
        ) {
            self.id = id
            self.role = role
            self.content = content
            self.timestamp = timestamp
        }
    }

    // MARK: - Bio Tone

    enum BioTone: String, CaseIterable, Identifiable {
        case witty = "Witty & Playful"
        case sincere = "Sincere & Genuine"
        case bold = "Bold & Confident"

        var id: String { rawValue }

        var icon: String {
            switch self {
            case .witty: return "face.smiling.inverse"
            case .sincere: return "heart.fill"
            case .bold: return "flame.fill"
            }
        }

        var description: String {
            switch self {
            case .witty: return "Humor-forward, clever wordplay"
            case .sincere: return "Heartfelt, authentic vibes"
            case .bold: return "Direct, high-energy, memorable"
            }
        }
    }

    // MARK: - Init

    init() {}

    // MARK: - Generate Bio

    func generateBio(
        interests: [String],
        tone: BioTone,
        platform: DatingPlatform
    ) async throws -> String {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        switch ServiceMode.current {
        case .production:
            // LAUNCH TODO: Call Firebase Cloud Function "generateBio"
            // POST /generateBio { interests, tone, platform }
            // Parse response JSON → String bio
            return try await generateMockBio(tone: tone)

        case .mock:
            return try await generateMockBio(tone: tone)
        }
    }

    private func generateMockBio(tone: BioTone) async throws -> String {
        try await Task.sleep(nanoseconds: 1_500_000_000)

        let bios: [BioTone: [String]] = [
            .witty: [
                "6'1\" but my personality is what really stands out at crowded bars. Dog dad. Will debate pizza toppings with dangerous passion. Looking for someone who laughs at their own jokes (because I will too).",
                "Pro at finding the best brunch spots within a 5-mile radius. My ideal Sunday: farmers market → cook together → lose at Scrabble. Fluent in sarcasm and movie quotes.",
                "If we match, I promise excellent playlist curation and slightly above-average cooking. Trade offer: I bring the adventure, you bring the snacks."
            ],
            .sincere: [
                "Looking for something real in a swipe-right world. I love exploring new neighborhoods, weekend hikes, and those deep late-night conversations. Looking for someone kind, curious, and down to build something great together.",
                "I believe in showing up fully — for the people I care about and the things I'm passionate about. Love traveling, live music, and learning to be a better cook. Here to find my person.",
                "Moved here recently and fell in love with the city. Now looking for someone to share the best parts with. I value honesty, laughter, and being able to sit in comfortable silence."
            ],
            .bold: [
                "I'm the one your friends will say \"he seems fun\" about. Entrepreneur by day, amateur chef by night. I don't waste time — if we vibe, let's grab drinks this week.",
                "Main character energy only. Building cool things, staying active, and always planning the next trip. Looking for someone who matches my energy and isn't afraid to make the first move.",
                "Life's too short for boring dates — I'm planning the best one you've had this year. Bring your A-game and an appetite. Let's see if we've got chemistry."
            ]
        ]

        let options = bios[tone] ?? bios[.witty] ?? []
        DesignSystem.Haptics.success()
        return options.randomElement() ?? "Your profile is looking great!"
    }

    // MARK: - Generate Opening Lines

    func generateOpeningLines(
        matchName: String,
        platform: DatingPlatform,
        context: String? = nil
    ) async throws -> [String] {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        switch ServiceMode.current {
        case .production:
            // LAUNCH TODO: Call Firebase Cloud Function "generateOpeningLines"
            // POST /generateOpeningLines { matchName, platform, context }
            return try await generateMockOpeningLines(matchName: matchName, context: context)

        case .mock:
            return try await generateMockOpeningLines(matchName: matchName, context: context)
        }
    }

    private func generateMockOpeningLines(matchName: String, context: String?) async throws -> [String] {
        try await Task.sleep(nanoseconds: 1_200_000_000)

        let lines: [[String]] = [
            [
                "Hey \(matchName)! I noticed \(context ?? "your profile") — gotta say, you've got great taste. What's the best thing you've done this week?",
                "Okay \(matchName), I need you to settle a debate: pineapple on pizza — yay or nay? (Your answer determines everything 😄)",
                "I had a really good opening line prepared, \(matchName), but then I saw your profile and forgot it. So let's just skip to the good part — what's your go-to weekend plan?"
            ],
            [
                "\(matchName)! Your \(context ?? "vibe") caught my eye. Quick question: if you could teleport anywhere right now, where are we going?",
                "Alright \(matchName), I'm gonna be honest — I swiped right faster than I'd like to admit. So what's the one thing I should know about you?",
                "Hey \(matchName) 👋 I'm not great at small talk but excellent at planning fun dates. Want to skip straight to that part?"
            ]
        ]

        DesignSystem.Haptics.success()
        return lines.randomElement() ?? lines[0]
    }

    // MARK: - Hinge Prompts

    func generateHingePrompts() async throws -> [(prompt: String, answer: String)] {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        switch ServiceMode.current {
        case .production:
            // LAUNCH TODO: Call Firebase Cloud Function "generateHingePrompts"
            return try await generateMockHingePrompts()

        case .mock:
            return try await generateMockHingePrompts()
        }
    }

    private func generateMockHingePrompts() async throws -> [(prompt: String, answer: String)] {
        try await Task.sleep(nanoseconds: 1_000_000_000)

        let prompts: [(String, String)] = [
            ("The way to win me over is", "Show up with curiosity and a willingness to try that weird restaurant I've been eyeing"),
            ("I'm looking for", "Someone who can match my energy at a concert AND my calm on a lazy Sunday morning"),
            ("A life goal of mine", "To visit every continent and have a local friend on each one to show me the hidden gems"),
            ("My most irrational fear", "That one day autocorrect will ruin the perfect text at the worst possible moment"),
            ("I bet you can't", "Name a better brunch spot than the one I found last weekend. Challenge accepted?"),
            ("Green flags I look for", "You send memes at appropriate times, tip well, and can laugh at yourself")
        ]

        DesignSystem.Haptics.success()
        return Array(prompts.shuffled().prefix(3))
    }

    // MARK: - Conversation Reply

    func suggestReply(
        to message: String,
        matchName: String,
        conversationContext: String? = nil
    ) async throws -> [String] {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        switch ServiceMode.current {
        case .production:
            // LAUNCH TODO: Call Firebase Cloud Function "suggestReply"
            // POST /suggestReply { message, matchName, conversationContext }
            return try await generateMockReplies()

        case .mock:
            return try await generateMockReplies()
        }
    }

    private func generateMockReplies() async throws -> [String] {
        try await Task.sleep(nanoseconds: 800_000_000)

        let replies = [
            "That's awesome! I've actually been wanting to try that. When are you free this week?",
            "Haha no way 😂 okay you're already cooler than most people I've matched with. Tell me more!",
            "Love that energy. I feel like we'd get along really well — want to find out over coffee?"
        ]

        DesignSystem.Haptics.success()
        return replies
    }
}
